### PR TITLE
User authentication and sign-up

### PR DESCRIPTION
**Pull request by noahboegli**
Updates here: https://dgheig.atlassian.net/jira/software/c/projects/P2/boards/5?modal=detail&selectedIssue=P2-14

### CODE REVIEW

File: app/backend/src/DataFixtures/UserFixture.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Persistence\ObjectManager;
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+
+class UserFixture extends Fixture
+{
+    private UserPasswordHasherInterface $passwordHasher;
+
+    /**
+     * Auto-wiring fuckers !!
```
Comment (dgheig) [inline app/backend/src/DataFixtures/UserFixture.php:15]: @noahboegli Please avoid insults

File: app/backend/src/DataFixtures/UserFixture.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+     * @param \Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface $passwordHasher
+     */
+    public function __construct(UserPasswordHasherInterface $passwordHasher)
+    {
+        $this->passwordHasher = $passwordHasher;
+    }
+
+    public function load(ObjectManager $manager)
```
Comment (dgheig) [inline app/backend/src/DataFixtures/UserFixture.php:23]: I will let devs evaluate the code but for the login part, I would like @heigvdgachetje to take a look at it from a pure security perspective, at least at the end of the dev if not during.

File: app/backend/.env [hunk 3]
```diff
@@ -31,4 +31,9 @@ DATABASE_URL="mysql://db:db@localhost:3306/db?serverVersion=10.3&charset=utf8mb4
 CORS_ALLOW_ORIGIN='^https?:\/\/(localhost|127\.0\.0\.1|(.+\.)?our-app\.local)(:[0-9]+)?$'
 ###< nelmio/cors-bundle ###
 
-VAR_DIR=/var/symfony
\ No newline at end of file
+VAR_DIR=/var/symfony
+###> lexik/jwt-authentication-bundle ###
+JWT_SECRET_KEY=%kernel.project_dir%/config/jwt/private.pem
```
Comment (dgheig) [inline app/backend/.env:36]: Can you keep tracks of the documentation you use and the steps we would need to replace/set our own token/crediential for the application? I think we can have a dedicated documentation file for that, so that any dev would be able to take over your work if needed in the future.

**Review by dgheig — COMMENTED** [review]
@noahboegli I know this is still a draft, I just added a few comment hoping it would spare you time when doing it now instead of later

File: app/backend/src/DataFixtures/UserFixture.php (hunk 2 above)
Comment (noahboegli) [reply app/backend/src/DataFixtures/UserFixture.php:23]: Yes. That is temporary until we do have a signup function (or at least user creation logic), which should happen before the merge.

File: app/backend/.env (hunk 3 above)
Comment (noahboegli) [reply app/backend/.env:36]: Yes. That is also planned to be added before the merge. All the details of what remains to be done are in the Jira Issue ;)

File: app/backend/src/DataFixtures/UserFixture.php (hunk 2 above)
Comment (noahboegli) [reply app/backend/src/DataFixtures/UserFixture.php:23]: And to be fair, the whole login part is implemented as per documented by symfony. Unless I have made a big big mistake when implementing it (which should be spotted during the review for this merge request) we shouldn't have to worry about the deep security aspects (or some other big companies out here should start to worry too :P)

**Comment by noahboegli** [comment]
Only docs remain, but that will be done another day.

In the mean time, let's enjoy our code coverage:

```
Code Coverage Report:       
  2022-04-15 12:08:56       
                            
 Summary:                   
  Classes: 100.00% (17/17)  
  Methods: 100.00% (121/121)
  Lines:   100.00% (214/214)
```

**Comment by dgheig** [comment]
@noahboegli I tried a quick run to check it and received a 401 HTTP error code.
If you are already aware then you can simply ignore this message.
Thank you for the changes

**Comment by noahboegli** [comment]
@dgheig Can you ellaborate further? There should be no 401 except when you access a protected page without a valid JWT. Note that Swagger's URL is now only `/docs`.

**Comment by noahboegli** [comment]
Aight @nigelmann @dgheig @Kitsunemma @heigvdgachetje 

This PR is now finally ready to merge. Thanks to @nigelmann for all the work on the unit testing capabilities enhancement and the serialization groups. 

@HugoMendes98 We'll catch each other either this week or next week to see how all of that can be integrated on your end.

I've added a small document called `auth.md` with a few points. Otherwise everything is either documented in swagger, the code itself or self-document by the usage of best-practices when implementing auth in Symfony.

@Kitsunemma Everything is tested: New entities, new controllers, user authentication & the firewall making sure that the user as the correct permissions. We've increased the code coverage to 100% with this PR.  

I'm available for feedback, changes or more documentation, but what I'd like the most is a positive review ;) 

Cheers

File: app/backend/src/DataFixtures/UserFixture.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+use Doctrine\Bundle\FixturesBundle\Fixture;
+use Doctrine\Persistence\ObjectManager;
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+
+class UserFixture extends Fixture
+{
+    private UserPasswordHasherInterface $passwordHasher;
+
+    /**
+     * @param \Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface $passwordHasher
```
Comment (nigelmann) [inline app/backend/src/DataFixtures/UserFixture.php:15]: Is this comment necessary ? the property is already typed below, to me this seems useless clutter

**Review by nigelmann — COMMENTED** [review]
I have toke the liberty to change the way back-end tests are run see in the workflow. The tests had not run which is a big no no for a pull request. I have also corrected a linting issue.

In order to have the CI running correctly I had to add a environment variable to define the SQLite database path, this way it can be overridden in the GitHub actions.

Anyway @noahboegli thanks for the effort !

File: app/backend/src/DataFixtures/UserFixture.php (hunk 4 above)
Comment (noahboegli) [reply app/backend/src/DataFixtures/UserFixture.php:15]: You don't like comments do you ? 

We can keep it or remove it. Since it was generated automatically there must be others. I suggest we take some time at the end of the next sprint to remove all of them, or add them everywhere, as you wish.

File: app/backend/tests/Unit/ApiTestTrait.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->client = static::createClient();
+        $this->databaseTool = static::getContainer()->get(DatabaseToolCollection::class)->get();
+        $this->databaseTool->loadFixtures([UserFixture::class]);
+    }
+
+    public function authenticatedRequest(string $method, string $path, array $options = []): ResponseInterface
```
[the remark this answers is not in the document]
Comment (noahboegli) [reply app/backend/tests/Unit/ApiTestTrait.php:27]: It's tests we don't really care 
Especially since we're using a dedicated database.

However it potentially hides issues with the authorisation mechanism since we request a new token each time. 

I'd suggest the same as you.

File: app/backend/src/DataFixtures/UserFixture.php (hunk 4 above)
Comment (nigelmann) [reply app/backend/src/DataFixtures/UserFixture.php:15]: Sounds good to me !

File: app/backend/tests/Unit/ApiTestTrait.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (nigelmann) [reply app/backend/tests/Unit/ApiTestTrait.php:27]: Alright !

**Review by nigelmann — APPROVED** [review]

**Comment by dgheig** [comment]
Hi @noahboegli  and thank you for the development.

I made some tests and here are some remarks

auth1
I was expecting to be logged-in on user creation without having to do a second query. Also, we would need stricter form check.

auth2
I know some/most/all of those "leaks" are only supposed to be used in dev, but as I only follow the README I am not aware if it is the case or not here, if I am really running a dev environment instead of a production one. When testing, functional testers should see the product in its production version.

auth3


Also, I am not confident in returning the token in the request body. I have done some reading myself afterwards and apparently this is indeed not recommended: [How to securely store JWTs](https://blog.logrocket.com/jwt-authentication-best-practices/#store-jwts-securely)
There are many other links, forums etc out there.
Could you take this recommendation (and/or any other found online) into account? If you have your own reason considering other sources I would be happy to read them.
Another issue with that is that the `/docs` does not allow the token to be added in queries (or am I missing something), and even if it could:
* It would be cluttersome
* and unsafe


@nigelmann Thank you for the fixes in the CI process. It would have been nice to have a different branch merged into this one through a proper PR for the tracing a validation, can you do that for the next CI changes?

**Comment by noahboegli** [comment]
@dgheig 

Debug messages in error / Dev env: We could add some documents to specify that the app is by default in dev mode, but for now we do not have a configuration ready for a production mode at the moment. And to answer your question: The messages won't be here in "production" mode, and by default, the Swagger UI is disabled in production mode.

E-mail validation: I may have forgotten to add a validator :P, I'll fix that.

Username: Initially I used a username, but I was told this was not in the spec so I reverted to an e-mail address. Tbh I'd be kinda pissed if now we have to revert back to a username.

JWT storage/Retrieval: There are millions of Symfony application using this method to retrieve the JWT. It's as per the specification. I don't see a problem with that. 

User creation/JWT retrieval: Creating a user should not automatically authenticate it (there are multiple use case against that). And since the user creation will be completely wrapped in the front-end, we can make it invisible for the user.

To summarize, here is what I will change
- Email format validation
- Documentation about being in dev mode. 

For all the other points, if you want to push changes we will have to discuss them further before. However I wonder if that should not be included in another PR/issue. Otherwise we will be pushing the release and the front-end integration further and further.

**Comment by dgheig** [comment]
Since the frontend will implement the login based on what is received, this need to be clarified first. Username can be left aside for the moment, this was just in prevision when managing friend list and groups.

According to my readings, JWT may be returned in the payload but for non web-browser usages.
I read 5-6 different blogs/forums and took a quick look to some others and none was recommending this. If you have other sources, then we can discuss it, otherwise this would need to be corrected.

For the automatic login, okay. Even if we don't implement email validation it would be half the job done already.

**Comment by noahboegli** [comment]
The JWT token retrieval is not described in the RFC (which is a bit of a shame tbh). 

While I agree with you that returning it in a cookie would be somewhat safer when it comes to token stealing. However, making it so would, as you mentioned, lock the authentication mechanism to a web-browser based client. This isn't ideal as it would be totally acceptable to think that our application could be ported to mobile devices. 

Also, I think we should discuss with @HugoMendes98 because he will be the first one to be impacted by such changes. I don't know how flexible our front-end stack is when it comes to token retrieval. 

Finally, returning the JWT token through the response's body is not a novelty. It is the default behaviour for a reason: It's the industry standard. Event OAuth uses this method (although you do have some extended mechanisms such as the PKCE variant to prevent such token hi-jacking). 

I may look stubborn, but I still think that storing it in the cookies isn't a gain we will benefit from. It's approximately 2 hours of additional work (maybe more since the tests might have to be re-worked). And that's without speaking of the chances with the password policy and the additional tests to be written. Yes it will add another layer of security, but given all the 0-days that we've seen so far when it comes to web browsers, I'm not sure it's worth the effort and the deviance from the industry standard. The effort should instead be put on ensuring that we store it correctly on the front-end side.

If you still disagree with that, then no problem I'll do it, but based on my experience, we should not.

In the mean time I'll push the changes discussed above (e-mail, password & docs for the testing environment) this afternoon.

**Comment by noahboegli** [comment]
With the new code & tests

```
Code Coverage Report:       
  2022-04-21 15:22:21       
                            
 Summary:                   
  Classes: 100.00% (19/19)  
  Methods: 100.00% (126/126)
  Lines:   100.00% (243/243)
```

**Comment by dgheig** [comment]
@noahboegli first, thank you for the quick changes.

I just want to point out 2 things:
- when i spoke of web-browser, I did not say that using cookies will bind it to this single usage. Accessing cookies can be done from any http client but in a less seamlessly manner.
- You mentionned it: the frontend storage, and there is no other that cookies that is considered safe, otherwise I wouldn't mind (at least not that much)
(And having a potential 0-days is not a good reason to give up on security)

As far as I know, those tokens are sent back this way when the frontend is not the destination but the backend

(Note for myself, TODO: Add github issues link found addressing this concern)

I am quit surprise that it would take 2h to change it?

**Comment by noahboegli** [comment]
Changing it would take that much time because I have to research how to do it properly, then I actually have to implement it (it's not the standard behavior so overriding it does require some time to ensure it's done properly...) 

And then we have to update the tests that are designed around the way it works for now, update the docs, update Swagger to ensure it will work properly with a cookie-based token. 

But I have another point that worries me, I'm not sure we're understanding each others points clearly, I wouldn't want to decide that based a misunderstanding between us :P 

Let's arrange a meeting to discuss about that this weekend. 
@others: We'll provide the minutes here afterwards.

**Comment by dgheig** [comment]
@noahboegli Here are some issues I found on one php lib as mentioned earlier. This may not be the lib we use but it can still be of some help. The purpose here is to have a common base on what we will discuss:

* [Issues · lexik/LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle/issues?page=4&q=Security)
* [Correct way to login · Issue #873 · lexik/LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle/issues/873)
* [Support setting refresh token in cookie by Jayfrown · Pull Request #274 · markitosgv/JWTRefreshTokenBundle](https://github.com/markitosgv/JWTRefreshTokenBundle/pull/274)
* [confused authentication · Issue #998 · lexik/LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle/issues/998)
* [httponly cookie expiration event handling · Issue #660 · lexik/LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle/issues/660)
* [Can you provide or anyone a map or recomendation to use this bundle with react and vue in client as SPA with no Symfony on client Side? · Issue #695 · lexik/LexikJWTAuthenticationBundle](https://github.com/lexik/LexikJWTAuthenticationBundle/issues/695)

There were more others and it would certainly be more useful ones on the repository of the lib we use.

File: app/backend/.env.test [hunk 6]
```diff
@@ -4,3 +4,5 @@ APP_SECRET='$ecretf0rt3st'
 SYMFONY_DEPRECATIONS_HELPER=999999
 PANTHER_APP_ENV=panther
 PANTHER_ERROR_SCREENSHOT_DIR=./var/error-screenshots
+SQLITE_PATH="$VAR_DIR/data.db"
+DATABASE_URL="sqlite:///$SQLITE_PATH"
```
Comment (heigvdgachetje) [inline app/backend/.env.test:8]: No new line at the end of the file

File: app/backend/config/routes/jwt.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+authentication_token:
+  path: /token/get
+  methods: ["POST"]
+gesdinet_jwt_refresh_token:
+  path: /token/refresh
```
Comment (heigvdgachetje) [inline app/backend/config/routes/jwt.yaml:5]: No new line at the end of file

**Review by dgheig — APPROVED** [review]
We will go forward as it is.
But be sure to have a singleton where to handle the authentication on both end in case this will be changed in the future

File: app/backend/scripts/setup.sh [hunk 8]
```diff
@@ -4,4 +4,4 @@ composer install
 ./bin/console doctrine:database:drop --force --if-exists
 ./bin/console doctrine:database:create --no-interaction
 ./bin/console doctrine:migration:migrate --no-interaction
-./bin/console doctrine:fixtures:load --no-interaction
+./bin/console doctrine:fixtures:load --no-interaction
```
Comment (heigvdgachetje) [inline app/backend/scripts/setup.sh:7]: No new line at the end of the file

File: app/backend/src/DataPersisters/AppUserDataPersister.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public function persist($data, array $context = [])
+    {
+        $password = $data->getPassword();
+
+        $minDigits = 1;
+        $minSpecial = 1;
+        $minUppercase = 1;
+        $minLength = 8;
```
Comment (heigvdgachetje) [inline app/backend/src/DataPersisters/AppUserDataPersister.php:34]: Shouldn't we add at least a lowercase too?

File: app/backend/src/Decorators/JwtGetDecorator.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                                          "description" => "Get JWT token",
+                                          "content" => [
+                                              "application/json" => [
+                                                  "schema" => [
+                                                      "\$ref" => "#/components/schemas/Token",
+                                                  ],
+                                              ],
+                                          ],
+                                      ],
+                                  ],
```
Comment (heigvdgachetje) [inline app/backend/src/Decorators/JwtGetDecorator.php:72]: One too many tabulation

File: app/backend/src/Decorators/JwtGetDecorator.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                summary:     "Gets a new JWT token.",
+                requestBody: new Model\RequestBody(
+                    description: "Generate new JWT Token",
+                    content:     new ArrayObject([
+                                                                       "application/json" => [
+                                                                           "schema" => [
+                                                                               "\$ref" => "#/components/schemas/Credentials",
+                                                                           ],
+                                                                       ],
+                                                                   ]),
```
Comment (heigvdgachetje) [inline app/backend/src/Decorators/JwtGetDecorator.php:82]: Two extra tabulation

**Review by heigvdgachetje — APPROVED** [review]
I just pointed out some linting that could be better but should not prevent this PR to be merged.

File: app/backend/src/Decorators/JwtGetDecorator.php (hunk 11 above)
Comment (noahboegli) [reply app/backend/src/Decorators/JwtGetDecorator.php]: This has been fixed in 8206ff91c941a46929608d33dc3d5443b3bb16bf

File: app/backend/src/Decorators/JwtGetDecorator.php (hunk 10 above)
Comment (noahboegli) [reply app/backend/src/Decorators/JwtGetDecorator.php]: This has been fixed in 8206ff91c941a46929608d33dc3d5443b3bb16bf

File: app/backend/src/DataPersisters/AppUserDataPersister.php (hunk 9 above)
Comment (noahboegli) [reply app/backend/src/DataPersisters/AppUserDataPersister.php]: This has been fixed in 617449981f95bde4c4519a05c3d6ee0693fc92e8

File: app/backend/scripts/setup.sh (hunk 8 above)
Comment (noahboegli) [reply app/backend/scripts/setup.sh]: This has been fixed in f7b3e00ce81226a571aa2f6cbbc446fef90529c9

File: app/backend/config/routes/jwt.yaml (hunk 7 above)
Comment (noahboegli) [reply app/backend/config/routes/jwt.yaml]: This has been fixed in f7b3e00ce81226a571aa2f6cbbc446fef90529c9

File: app/backend/.env.test (hunk 6 above)
Comment (noahboegli) [reply app/backend/.env.test]: This has been fixed in f7b3e00ce81226a571aa2f6cbbc446fef90529c9

**Comment by noahboegli** [comment]
@heigvdgachetje  Your changes have been implemented in the 3 last commits. :) 

@Kitsunemma Could you please take some time to review it today? That way we could merge it before today's session. Thanks !

**Review by Kitsunemma — APPROVED** [review]
